README
NAME
    Data::Validate::Domain - domain validation methods

SYNOPSIS
      use Data::Validate::Domain qw(is_domain);
 
      # as a function 
      my $test = is_domain($suspect);
      die "$test is not a domain" unless defined $test;

      or

      my $test = is_domain($suspect,\%options);
      die "$test is not a domain" unless defined $test;
  
      # or as an object
      my $v = Data::Validate::Domain->new(%options);

      my $test = $v->is_domain($suspect);
      die "$test is not a domain" unless defined $test;

DESCRIPTION
    This module collects domain validation routines to make input
    validation, and untainting easier and more readable.

    All functions return an untainted value if the test passes, and undef if
    it fails. This means that you should always check for a defined status
    explicitly. Don't assume the return will be true. (e.g.
    is_username('0'))

    The value to test is always the first (and often only) argument.

FUNCTIONS
    new - constructor for OO usage
          $obj = Data::Validate::Domain->new();

          my %options = (
                        domain_allow_single_label => 1,
                        domain_private_tld => {
                                'privatetld1 '   =>      1,
                                'privatetld2'    =>      1,
                        }
          );

          or

          my %options = (
                        domain_allow_single_label => 1,
                        domain_private_tld        => qr /^(?:privatetld1|privatetld2)$/,
          );

          $obj = Data::Validate::Domain->new(%options);

        *Description*
            Returns a Data::Validator::Domain object. This lets you access
            all the validator function calls as methods without importing
            them into your namespace or using the clumsy
            Data::Validate::Domain::function_name() format.

        *Options*

            domain_allow_single_label
                By default is_domain will fail if you ask it to verify a
                domain that only has a single label i.e. 'neely.cx' is good,
                but 'com' would fail. If you set this option to a true value
                then is_domain will allow single label domains through. This
                is most likely to be useful in combination with
                domain_private_tld

            domain_private_tld
                By default is_domain requires all domains to have a valid
                TLD (i.e. com, net, org, uk, etc), this is verified using
                the Net::Domain::TLD module. This behavior can be extended
                in two different ways. Either a hash reference can be
                supplied keyed by the additional TLD's, or you can supply a
                precompiled regular expression.

                NOTE: The TLD is normalized to the lower case form prior to
                the check being done. This is done only for the TLD check,
                and does not alter the output in any way.

                        The hash reference example:     

                                domain_private_tld => {
                                        'privatetld1 '   =>      1,
                                        'privatetld2'    =>      1,
                                }

                        The precompiled regualar expression example:

                                domain_private_tld        => qr /^(?:privatetld1|privatetld2)$/,

        *Returns*
            Returns a Data::Validate::Domain object

    is_domain - does the value look like a domain name?
          is_domain($value);
          or
          $obj->is_domain($value);
          or
          is_domain($value,\%options);
          or
          $obj->is_domain($value,\%options);

        *Description*
            Returns the untainted domain name if the test value appears to
            be a well-formed domain name.

            Note: See new for list of options and how those alter the
            behavior of this funciton.

        *Arguments*

            $value
                The potential domain to test.

        *Returns*
            Returns the untainted domain on success, undef on failure.

        *Notes, Exceptions, & Bugs*
            The function does not make any attempt to check whether a domain
            actually exists. It only looks to see that the format is
            appropriate.

            A dotted quad (such as 127.0.0.1) is not considered a domain and
            will return false. See Data::Validate::IP(3) for IP Validation.

            Performs a lookup via Net::Domain::TLD to verify that the TLD is
            valid for this domain.

            Does not consider "domain.com." a valid format.

        *From RFC 952*
               A "name" (Net, Host, Gateway, or Domain name) is a text string up
               to 24 characters drawn from the alphabet (A-Z), digits (0-9), minus
               sign (-), and period (.).  Note that periods are only allowed when
               they serve to delimit components of "domain style names".

               No blank or space characters are permitted as part of a
               name. No distinction is made between upper and lower case.  The first
               character must be an alpha character [Relaxed in RFC 1123] .  The last 
               character must not be a minus sign or period.

        *From RFC 1035*
                labels          63 octets or less
                names           255 octets or less

                [snip] limit the label to 63 octets or less.

                To simplify implementations, the total length of a domain name (i.e.,
                label octets and label length octets) is restricted to 255 octets or
                less.

        *From RFC 1123*
                One aspect of host name syntax is hereby changed: the
                restriction on the first character is relaxed to allow either a
                letter or a digit.  Host software MUST support this more liberal
                syntax.

                Host software MUST handle host names of up to 63 characters and
                SHOULD handle host names of up to 255 characters.

    is_hostname - does the value look like a hostname
          is_hostname($value);
          or
          $obj->is_hostname($value);
          or
          is_hostname($value,\%options);
          or
          $obj->is_hostname($value,\%options);

        *Description*
            Returns the untainted hostname if the test value appears to be a
            well-formed hostname.

            Note: See new for list of options and how those alter the
            behavior of this funciton.

        *Arguments*

            $value
                The potential hostname to test.

        *Returns*
            Returns the untainted hostname on success, undef on failure.

        *Notes, Exceptions, & Bugs*
            The function does not make any attempt to check whether a
            hostname actually exists. It only looks to see that the format
            is appropriate.

            Functions much like is_domain, except that it does not verify
            whether or not a valid TLD has been supplied and allows for
            there to only be a single component of the hostname (i.e www)

            Hostnames might or might not have a valid TLD attached.

    is_domain_label - does the value look like a domain label?
          is_domain_label($value);
          or
          $obj->is_domain_label($value);
          or
          is_domain_label($value,\%options);
          or
          $obj->is_domain_label($value,\%options);

        *Description*
            Returns the untainted domain label if the test value appears to
            be a well-formed domain label.

            Note: See new for list of options and how those alter the
            behavior of this funciton.

        *Arguments*

            $value
                The potential ip to test.

        *Returns*
            Returns the untainted domain label on success, undef on failure.

        *Notes, Exceptions, & Bugs*
            The function does not make any attempt to check whether a domain
            label actually exists. It only looks to see that the format is
            appropriate.

SEE ALSO
        [RFC 1034] [RFC 1035] [RFC 2181] [RFC 1123]

        Data::Validate(3)
        Data::Validate::IP(3)

AUTHOR
        Neil Neely <neil@neely.cx>.

ACKNOWLEDGEMENTS
        Thanks to Richard Sonnen <sonnen@richardsonnen.com> for writing the
        Data::Validate module.

        Thanks to Len Reed <lreed@levanta.com> for helping develop the
        options mechanism for Data::Validate modules.

COPYRIGHT AND LICENSE
        Copyright (c) 2005-2006 Neil Neely.

        This library is free software; you can redistribute it and/or modify
        it under the same terms as Perl itself, either Perl version 5.8.2
        or, at your option, any later version of Perl 5 you may have
        available.



Changes
{{$NEXT}}

0.15     2021-08-21
    - Reject domains with 254 or more characters. Domains are limited to 255
      octets, but in practice this means they're actually limited to 253
      characters. See
      https://devblogs.microsoft.com/oldnewthing/20120412-00/?p=7873 for a
      more detailed explanation of this. Implemented by Felipe Gaspar. Fixes
      GitHub #5. PR #13.


0.14     2016-08-21
    - Require Net::Domain::TLD 1.74. This fixes some failures seen on CPAN
      Testers, and if you're validating domains you really want the latest
      version of this module.

0.13     2016-08-21
    - Accept trailing dot on hostnames as well. Implemented by Greg
      Oschwald. GitHub #6.
    - The option domain_disable_tld_validation was added to allow disabling
      TLD validation. This may be useful if you need to check domains with new
      gTLDs that have not yet been added to Net::Domain::TLD. Implemented by Greg
      Oschwald. GitHub #8.

0.12     2016-06-05
    - Accept trailing dot on a domain name ("example.com."). This is an
      absolute name per RFCs 1034 & 1035, and will be accepted by any DNS
      resolution library or tool. Reported by Rob J. Epping.

0.11     2015-05-05
    - Don't allow non-ASCII digits in domain labels, RT #103121.

0.10     2010-12-29
    - Fixed bug 41033, no longer allows \n in domain_labels
        Thanks to marek.miska@netart.pl for reporting the bug

0.09     2008-02-18
    - Added new option of domain_allow_underscore to allow this module
      to be used for validating domainnames like _spf.neely.cx if the option is set
      Default behavior was not changed

0.08     2007-05-17
    - Ugh, forgot to add new tests to MANIFEST

0.07     2007-05-17
    - Added POD testing, and minor cleanup related to that

0.06     2007-05-16
    - Updated contact information to be neil@neely.cx, this is a purely cosmetic change

0.05     2006-12-04
    - Added Support for compiled regular expressions to the private_domain_tld option.
        - Thanks to Berhhard Graf <graf@augensalat.de> for the suggestion.

0.04     2006-12-01
    - Added support for options to alter behavior of is_domain with these new
      options is_domain will be able to work with private domains with custom TLD's.
        - Special Thanks to Len Reed <lreed@levanta.com> for not only
          suggesting the need for this but supplying several patches to help
          steer this solution in the right direction.  Also made several
          documentation improvements.
        - Additional thanks to Richard Sonnen <sonnen@richardsonnen.com> for
          helping flush out the details of how to utilize options in the
                  Data::Validate name space portably.

0.03     2005-11-03
    - Updated to support new procedural version of Net::Domain::TLD
        - Thanks to Ricardo Signes <rjbs@cpan.org> for letting me know about his new interface

0.02     2005-10-28
    - added is_hostname method (which is valid for either somehost or somehost.domain.com)
    - modified is_domain to verify TLD of domain supplied (Uses Net::Domain::TLD)
    - Fixed bug that allowed domain.com... to return the same results as domain.com
    - Will no longer consider domain.com. to be valid
        - Thanks to Smylers <smylers@donhost.co.uk> for the detailed bug
          report and suggested fixes.

0.01     2005-03-04
    - original version; created by h2xs 1.23 with options
        -AXn Data::Validate::Domain



Makefile.PL
use ExtUtils::MakeMaker;
# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Data::Validate::Domain',
    VERSION_FROM      => 'lib/Data/Validate/Domain.pm', # finds $VERSION
    LICENSE		=> 'perl',
    PREREQ_PM         => {
		                        Test::More	=>	0,
		                        Net::Domain::TLD => 1.62,
				}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Data/Validate/Domain.pm', # retrieve abstract from module
       AUTHOR         => 'Neil Neely <neil@neely.cx>') : ()),
    META_MERGE => {
        'meta-spec' => { version => 2 },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/neilneely/Data-Validate-Domain.git',
                web  => 'https://github.com/neilneely/Data-Validate-Domain',
            },
        },
    },
);


lib/Data/Validate/Domain.pm
package Data::Validate::Domain;

use strict;
use warnings;

our $VERSION = '0.16';

use Net::Domain::TLD 1.74 qw(tld_exists);

use Exporter qw( import );

## no critic (Modules::ProhibitAutomaticExportation)
our @EXPORT = qw(
    is_domain
    is_hostname
    is_domain_label
);

sub new {
    my $class = shift;

    return bless {@_}, ref($class) || $class;
}

# -------------------------------------------------------------------------------

sub is_domain {
    my ( $value, $opt ) = _maybe_oo(@_);

    my ( $hostname, $bits ) = _domain_labels( $value, $opt );

    return unless $bits;

    my $tld = $bits->[-1];

    # domain_allow_single_label set to true disables this check
    unless ( $opt->{domain_allow_single_label} ) {

        # All domains have more then 1 label (neely.cx good, com not good)
        return if @{$bits} < 2;
    }

    return $hostname if $opt->{domain_disable_tld_validation};

    # If the option to enable domain_private_tld is enabled
    # and a private domain is specified, then we return if that matches
    if ( exists $opt->{domain_private_tld}
        && ref( $opt->{domain_private_tld} ) ) {
        my $lc_tld = lc($tld);
        if ( ref( $opt->{domain_private_tld} ) eq 'HASH' ) {
            if ( exists $opt->{domain_private_tld}->{$lc_tld} ) {
                return $hostname;
            }
        }
        else {
            if ( $tld =~ $opt->{domain_private_tld} ) {
                return $hostname;
            }
        }
    }

    # Verify domain has a valid TLD
    return unless tld_exists($tld);

    return $hostname;
}

# -------------------------------------------------------------------------------

sub is_hostname {
    my ( $value, $opt ) = _maybe_oo(@_);

    my ($hostname) = _domain_labels( $value, $opt );

    # We do not verify TLD for hostnames, as hostname.subhost is a valid hostname

    return $hostname;
}

sub _domain_labels {
    my ( $value, $opt ) = @_;

    return unless defined($value);

    # FYI: DNS limits names to 255 octets, encoded to RDATA. Each label
    # includes a length-octet prefix; those length octets count against the
    # 255-octet maximum. The number of labels exceeds the number of dots by 1
    # (assuming no trailing dot), and the number of length octets exceeds
    # number of labels by 1 (since there’s always a trailing NUL octet). The
    # effective limit is thus 255 - 1 - 1, or 253. See
    # https://devblogs.microsoft.com/oldnewthing/20120412-00/?p=7873 for a
    # more detailed explanation of this.

    my $trailing_dot = $value =~ s/\.\z// ? q{.} : q{};

    my $encoded_length = ( $trailing_dot ? 1 : 2 ) + length($value);

    return if $encoded_length > 255;

    my @bits;
    foreach my $label ( split /\./, $value, -1 ) {
        my $bit = is_domain_label( $label, $opt );
        return unless defined $bit;
        push( @bits, $bit );
    }

    return unless @bits;

    return ( join( '.', @bits ) . $trailing_dot, \@bits );
}

sub is_domain_label {
    my ( $value, $opt ) = _maybe_oo(@_);

    return unless defined($value);

    # Fix Bug: 41033
    return if ( $value =~ /\n/ );

    # bail if we are dealing with more then just a hostname
    return if ( $value =~ /\./ );
    my $length = length($value);
    my $hostname;
    if ( $length == 1 ) {
        if ( $opt->{domain_allow_underscore} ) {
            ($hostname) = $value =~ /^([0-9A-Za-z\_])$/;
        }
        else {
            ($hostname) = $value =~ /^([0-9A-Za-z])$/;
        }
    }
    elsif ( $length > 1 && $length <= 63 ) {
        if ( $opt->{domain_allow_underscore} ) {
            ($hostname)
                = $value =~ /^([0-9A-Za-z\_][0-9A-Za-z\-\_]*[0-9A-Za-z])$/;
        }
        else {
            ($hostname)
                = $value =~ /^([0-9A-Za-z][0-9A-Za-z\-]*[0-9A-Za-z])$/;
        }
    }
    else {
        return;
    }
    return $hostname;
}

sub _maybe_oo {
    if ( ref $_[0] ) {
        return @_[ 1, 0 ];
    }
    else {
        return ( $_[0], ( defined $_[1] ? $_[1] : {} ) );
    }
}

1;

# ABSTRACT: Domain and host name validation

__END__

=pod

=for test_synopsis my ($suspect, %options);

=head1 SYNOPSIS

  use Data::Validate::Domain qw(is_domain);

  # as a function
  my $test = is_domain($suspect);
  die "$test is not a domain" unless $test;

  # or

  die "$test is not a domain" unless is_domain($suspect, \%options);

  # or as an object
  my $v = Data::Validate::Domain->new(%options);

  die "$test is not a domain" unless $v->is_domain($suspect);

=head1 DESCRIPTION

This module offers a few subroutines for validating domain and host names.

=head1 FUNCTIONS

All of the functions below are exported by default.

All of the functions return an untainted value on success and a false value
(C<undef> or an empty list) on failure. In scalar context, you should check
that the return value is defined, because something like
C<is_domain_label('0')> will return a defined but false value.

The value to test is always the first (and often only) argument.

Note that none of these functions test whether a domain or hostname is actually
resolvable or reachable.

=head2 Data::Validate::Domain->new()

This method constructs a validation object. It accepts the following arguments:

=over 4

=item * domain_allow_underscore

According to RFC underscores are forbidden in hostnames but not domain names.
By default C<is_domain()>, C<is_domain_label()>, and C<is_hostname()> will fail
if the value to be checked includes underscores. Setting this to a true value
will allow the use of underscores in all functions.

=item * domain_allow_single_label

By default C<is_domain()> will fail if you ask it to verify a domain that only
has a single label i.e. "neely.cx" is good, but "com" would fail. If you set
this option to a true value then C<is_domain()> will allow single label domains
through. This is most likely to be useful in combination with the
C<domain_private_tld> argument.

=item * domain_disable_tld_validation

Disables TLD validation for C<is_domain()>. This may be useful if you need to
check domains with new gTLDs that have not yet been added to
L<Net::Domain::TLD>.

=item * domain_private_tld

By default C<is_domain()> requires all domains to have a valid public TLD (i.e.
com, net, org, uk, etc). This is verified using the L<Net::Domain::TLD> module.
This behavior can be extended in two different ways. You can provide either a
hash reference where additional TLDs are keys or you can supply a regular
expression.

NOTE: The TLD is normalized to the lower case form prior to the check being
done. This is done only for the TLD check, and does not alter the output in any
way.

Hashref example:

  domain_private_tld => {
      privatetld1 => 1,
      privatetld2 => 1,
  }

Regular expression example:

 domain_private_tld => qr /^(?:privatetld1|privatetld2)$/,

=back

=head2 is_domain($domain, \%options)

This can be called as either a subroutine or a method. If called as a sub, you
can pass any of the arguments accepted by the constructor as options. If called
as a method, any additional options are ignored.

This returns the untainted domain name if the given C<$domain> is a valid
domain.

A dotted quad (such as 127.0.0.1) is not considered a domain and will return
false. See L<Data::Validate::IP> for IP Validation.

Per RFC 1035, this sub does accept a value ending in a single period (i.e.
"domain.com.") to be a valid domain. This is called an absolute domain name,
and should be properly resolved by any DNS tool (tested with C<dig>, C<ssh>,
and L<Net::DNS>).

=over 4

=item I<From RFC 952>

   A "name" (Net, Host, Gateway, or Domain name) is a text string up
   to 24 characters drawn from the alphabet (A-Z), digits (0-9), minus
   sign (-), and period (.). Note that periods are only allowed when
   they serve to delimit components of "domain style names".

   No blank or space characters are permitted as part of a
   name. No distinction is made between upper and lower case. The first
   character must be an alpha character [Relaxed in RFC 1123] . The last
   character must not be a minus sign or period.

=item I<From RFC 1035>

    labels          63 octets or less
    names           255 octets or less

    [snip] limit the label to 63 octets or less.

    To simplify implementations, the total length of a domain name (i.e.,
    label octets and label length octets) is restricted to 255 octets or
    less.

=item I<From RFC 1123>

    One aspect of host name syntax is hereby changed: the
    restriction on the first character is relaxed to allow either a
    letter or a digit. Host software MUST support this more liberal
    syntax.

    Host software MUST handle host names of up to 63 characters and
    SHOULD handle host names of up to 255 characters.

=back

=head2 is_hostname($hostname, \%options)

This can be called as either a subroutine or a method. If called as a sub, you
can pass any of the arguments accepted by the constructor as options. If called
as a method, any additional options are ignored.

This returns the untainted hostname if the given C<$hostname> is a valid
hostname.

Hostnames are not required to end in a valid TLD.

=head2 is_domain_label($label, \%options)

This can be called as either a subroutine or a method. If called as a sub, you
can pass any of the arguments accepted by the constructor as options. If called
as a method, any additional options are ignored.

This returns the untainted label if the given C<$label> is a valid label.

A domain label is simply a single piece of a domain or hostname. For example,
the "www.foo.com" hostname contains the labels "www", "foo", and "com".

=head1 SEE ALSO

B<[RFC 1034] [RFC 1035] [RFC 2181] [RFC 1123]>

=over 4

=item  L<Data::Validate>

=item  L<Data::Validate::IP>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to Richard Sonnen <F<sonnen@richardsonnen.com>> for writing the
Data::Validate module.

Thanks to Len Reed <F<lreed@levanta.com>> for helping develop the options
mechanism for Data::Validate modules.

=cut

